### PR TITLE
SCC-3757 - Hold Page UI and Authentication Redirect

### PR DESCRIPTION
**Pull request by dgcohen**
## Ticket:

- JIRA ticket [SCC-3757](https://newyorkpubliclibrary.atlassian.net/browse/SCC-3757)

## This PR does the following:

- Initializes the hold request page component UI with dummy data in the radio buttons
- Adds auth redirect (derived form account) in getServerSide props, let me know if there is logic from MyAccount that doesn't belong here.
- Some basic tests.
- Fixes styling on bib detail labels (based on Apoorva's input)

## Updates 10-22-24

- Add optional itemId param in fetchBib to retrieve a specific item.
- Add infrastructure for post request with dummy response data
- Add error notifcation
- Populated feedback form with item details

## How has this been tested?

- Added initial page component unit tests

## Accessibility concerns or updates

NA

### Checklist:

- [x] I updated the CHANGELOG with the appropriate information and JIRA ticket number (if applicable).
- [x] I have added relevant accessibility documentation for this pull request.
- [x] All new and existing tests passed.

[SCC-3757]: https://newyorkpubliclibrary.atlassian.net/browse/SCC-3757?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: pages/hold/request/[id].tsx [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+        <meta property="og:title" content={metadataTitle} key="og-title" />
+        <meta
+          property="og:site_name"
+          content={metadataTitle}
+          key="og-site-name"
+        />
+        <meta name="twitter:title" content={metadataTitle} key="tw-title" />
+        <title key="main-title">{metadataTitle}</title>
+      </Head>
+      <Layout isAuthenticated={isAuthenticated} activePage="search">
```
Comment (EdwinGuzman) [inline pages/hold/request/[id].tsx:74]: There should be a new `activePage` value for the hold pages. Based on the designs, the hold pages don't have the search bar.

File: pages/hold/request/[id].tsx [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+                    <br />
+                    476 Fifth Avenue (42nd St and Fifth Ave)
+                  </>
+                }
+                value="4"
+              />
+            </RadioGroup>
+          </FormField>
+          <ButtonGroup>
+            <Button id="holdRequestSubmit" type="submit" size="large">
```
Comment (EdwinGuzman) [inline pages/hold/request/[id].tsx:167]: This shouldn't be large

File: pages/hold/request/[id].tsx [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+        </List>
+        <Heading level="h3" size="heading4" mb="l">
+          Choose a pickup location
+        </Heading>
+        <Form
+          id="hold-request-form"
+          data-testid="hold-request-form"
+          // We are using a post request on hold requests when JS is disabled
+          method="post"
+          action={`${BASE_URL}/api/hold/request/${bib.id}-${item.id}`}
```
Comment (EdwinGuzman) [inline pages/hold/request/[id].tsx:106]: to be clear, this route will be created in a different PR, right?

File: pages/hold/request/[id].tsx [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+  isAuthenticated,
+}: HoldRequestPropsType) {
+  const metadataTitle = `Item Request | ${SITE_NAME}`
+
+  const bib = new Bib(discoveryBibResult)
+  const item = new Item(discoveryItemResult, bib)
+
+  const handleSubmit = (e) => {
+    e.preventDefault()
+    console.log("submitting form")
```
Comment (EdwinGuzman) [inline pages/hold/request/[id].tsx:59]: I think you can try to mock the error and success state in this PR. So this can display the error `Banner` on the first click and then go to the confirmation route on the second click. It'll be mocked submissions but it can help push the UI and userflow along.

**Review by EdwinGuzman — CHANGES_REQUESTED** [review]
Looks good but it should have the error/success state even if they're mocked for now.

File: pages/hold/request/[id].tsx (hunk 4 above)
Comment (EdwinGuzman) [reply pages/hold/request/[id].tsx:59]: And for the error state, you can pass a mocked message to the `FeedbackBox` component and open it when the "contact us" button (that looks like a link) is clicked.

File: pages/hold/request/[id].tsx (hunk 1 above)
Comment (dgcohen) [reply pages/hold/request/[id].tsx]: Good call, i added this because I saw "Search" was active in the subnav, but forgot that this adds the search bar implicitly. Added a new hold type.

File: pages/hold/request/[id].tsx (hunk 3 above)
Comment (dgcohen) [reply pages/hold/request/[id].tsx:108]: Correct, i can add the skeleton now but will be hard to test before server functions are done.

File: pages/hold/request/[id].tsx (hunk 4 above)
Comment (dgcohen) [reply pages/hold/request/[id].tsx:59]: Good suggestions, the confirmation page has its own ticket but I can add the error state UI.

Comment (EdwinGuzman) [reply pages/hold/request/[id].tsx:59]: The confirmation page UI is its own ticket, but for this work it's enough to go to the confirmation page, even if it's broken.

File: pages/hold/request/[id].tsx [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+
+  // If we end up not authenticated 3 times after redirecting to the login url, don't redirect.
+  if (redirectBasedOnNyplAccountRedirects && !isAuthenticated) {
+    res.setHeader(
+      "Set-Cookie",
+      `nyplAccountRedirects=${
+        redirectCount + 1
+      }; Max-Age=10; path=/; domain=.nypl.org;`
+    )
+    const redirect = getLoginRedirect(req, "/account")
```
[the remark this answers is not in the document]
Comment (charmingduchess) [reply pages/hold/request/[id].tsx:198]: Does this implementation redirect you to the account page after logging in?

File: pages/hold/request/[id].tsx [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+    return {
+      redirect: {
+        destination: redirect,
+        permanent: false,
+      },
+    }
+  }
+
+  try {
+    const patronId = patronTokenResponse.decodedPatron.sub
```
[the remark this answers is not in the document]
Comment (dgcohen) [reply pages/hold/request/[id].tsx:209]: We need the patron id for the locations fetch

File: pages/hold/request/[id].tsx [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+    const { patron } = await getPatronData(patronId)
+
+    // fetch bib and item
+    const [bibId, itemId] = id.split("-")
```
[the remark this answers is not in the document]
Comment (dgcohen) [reply pages/hold/request/[id].tsx:213]: Cool didn't realize that (this was modeled after DFE)

File: pages/hold/request/[id].tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (dgcohen) [reply pages/hold/request/[id].tsx:198]: Good catch, fixed

File: pages/hold/request/[id].tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (dgcohen) [reply pages/hold/request/[id].tsx:209]: But i was able to get rid of the call to getPatronData

**Review by charmingduchess — COMMENTED** [review]
This is looking great! Just a couple file structure notes.

One more thing- I just remembered the real reason behind the item id endpoint.

The Discovery API does not return the most updated availability for partner items. During the fetching of search results, we put too much strain on SCSB if we ask for information about every single item. So we just pass along the potentially stale availability data we have indexed (which is updated nightly). It remains potentially stale until the -itemid endpoint is hit, at which point we do actually ask SCSB if a record is in fact available. The idea is that in the case of an item that is unavailable, to not let a patron make a hold request that we know will fail. Again, not sure if this is logic that ever made it anywhere, or if we want to add some handling for that here.

In short, there is updated availability data that we might want to check before letting a hold request go through.

@nonword can you fill in any gaps in this explanation?

File: src/components/ItemTable/ItemAvailability.tsx [hunk 8]
```diff
@@ -100,15 +101,7 @@ const ItemAvailability = ({ item }: ItemAvailabilityProps) => {
...
-          sx={{
-            display: "inline",
-            fontWeight: "inherit",
-            fontSize: "inherit",
-            p: 0,
-            height: "auto",
-            textAlign: "left",
-            minHeight: "auto",
-          }}
+          sx={BUTTON_LINK_STYLES}
```
[the remark this answers is not in the document]
Comment (dgcohen) [reply src/components/ItemTable/ItemAvailability.tsx:104]: Edwin added that comment in the advanced search page but I believe it should always be rendered to control the focus for accessibility. Just amended the comment.

File: pages/api/hold/request/[id].ts [hunk 9]
```diff
@@ -0,0 +1,28 @@
+import type { NextApiRequest, NextApiResponse } from "next"
+
+/**
+ * Default API route handler for Hold requests
+ * Calls a helper function that maps the query params object to a SearchParams object
```
Comment (EdwinGuzman) [inline pages/api/hold/request/[id].ts:5]: This isn't true but I know this is just the stub handler function.

File: pages/hold/request/[id].tsx [hunk 10]
```diff
@@ -0,0 +1,340 @@
...
+      setFormPosting(true)
+      const response = await fetch(`${BASE_URL}/api/hold/request/${holdId}`, {
+        method: "POST",
+        // TODO: serialize form data
+        body: "",
+      })
+      const responseJson = await response.json()
+
+      if (response.status !== 200) {
+        console.error("Error in hold  request api response", responseJson.error)
```
Comment (EdwinGuzman) [inline pages/hold/request/[id].tsx:95]: Prefix the string with `"HoldRequestPage: Error in..."` so it's easier to debug. Nitpick: extra space after "hold" and API instead of api.

File: pages/hold/request/[id].tsx [hunk 11]
```diff
@@ -71,7 +127,44 @@ export default function HoldRequestPage({
         <meta name="twitter:title" content={metadataTitle} key="tw-title" />
         <title key="main-title">{metadataTitle}</title>
       </Head>
-      <Layout isAuthenticated={isAuthenticated} activePage="search">
+      <Layout isAuthenticated={isAuthenticated} activePage="hold">
+        {/* Always render the wrapper element that will display the
```
[the remark this answers is not in the document]
Comment (EdwinGuzman) [reply pages/hold/request/[id].tsx:129]: For screenreaders to read the dynamic alert text.

File: pages/hold/request/[id].tsx [hunk 12]
```diff
@@ -0,0 +1,340 @@
...
+          content={metadataTitle}
+          key="og-site-name"
+        />
+        <meta name="twitter:title" content={metadataTitle} key="tw-title" />
+        <title key="main-title">{metadataTitle}</title>
+      </Head>
+      <Layout isAuthenticated={isAuthenticated} activePage="hold">
+        {/* Always render the wrapper element that will display the
+          dynamically rendered notification for focus management */}
+        <Box tabIndex={-1} ref={notificationRef}>
```
Comment (EdwinGuzman) [inline pages/hold/request/[id].tsx:131]: Nit: rename as `bannerRef` to be slightly more concise, or `bannerContainerRef`?

File: pages/hold/request/[id].tsx [hunk 13]
```diff
@@ -0,0 +1,340 @@
...
+          {alert && (
+            <Banner
+              type="negative"
+              heading="Request failed"
+              data-testid="hold-request-error"
+              sx={{ a: { color: "ui.link.primary" } }}
```
Comment (EdwinGuzman) [inline pages/hold/request/[id].tsx:137]: This should use the default error color for the link. CC @charmingduchess 
Check the documentation in Storybook and the "Design" tab: https://nypl.github.io/nypl-design-system/reservoir/v3/?path=/story/components-feedback-banner--with-controls 

Screenshot 2024-10-22 at 5 12 39 PM
Screenshot 2024-10-22 at 5 13 08 PM

File: pages/api/bib/[id].tsx [hunk 14]
```diff
@@ -12,9 +12,16 @@ import { PATHS, BASE_URL } from "../../../src/config/constants"
...
  */
 async function handler(req: NextApiRequest, res: NextApiResponse) {
-  const id = req.query.id as string
+  let bibId = req.query.id as string
+  let itemId: string | undefined
+
+  // If the id contains a hyphen, the id contains a bibId and an itemId
+  if (bibId.includes("-")) {
+    bibId = bibId.split("-")[0]
+    itemId = bibId.split("-")[1]
```
Comment (EdwinGuzman) [inline pages/api/bib/[id].tsx:21]: Can you do `[bibId, itemId] = id.split("-")` as you did in a different file?

File: pages/hold/request/[id].tsx [hunk 15]
```diff
@@ -0,0 +1,340 @@
...
+  }
+
+  try {
+    const patronId = patronTokenResponse?.decodedPatron?.sub
+
+    // fetch bib and item
+    const [bibId, itemId] = id.split("-")
+
+    if (!itemId) {
+      throw new Error("No item id in url")
```
Comment (EdwinGuzman) [inline pages/hold/request/[id].tsx:266]: I think at the root, this should be a 404 error and should redirect to that page.

Comment (EdwinGuzman) [reply pages/hold/request/[id].tsx:266]: It's user error, not an application error.

Comment (EdwinGuzman) [reply pages/hold/request/[id].tsx:266]: Ah, this is wrapped in a try/catch and does redirect. Great. I commented too early. But for the error on 324, that might not be a 404 but maybe it is a 400?

**Review by EdwinGuzman — COMMENTED** [review]
The update with the error banner works well, and added comments on this last round.

File: pages/hold/request/[id].tsx
[the remark this answers is not in the document]
Comment (dgcohen) [reply pages/hold/request/[id].tsx:1]: Done!

File: src/components/ItemTable/ItemAvailability.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (dgcohen) [reply src/components/ItemTable/ItemAvailability.tsx]: I ended up scrapping this, since i was just trying to cut down duplication since we override the button styles to match the DS link styles in one other place. This should be addressed on the DS level, will add TODO comments and bring it up in the next DS meeting.

File: pages/hold/request/[id].tsx (hunk 12 above)
Comment (dgcohen) [reply pages/hold/request/[id].tsx]: Done!

File: pages/hold/request/[id].tsx (hunk 13 above)
Comment (dgcohen) [reply pages/hold/request/[id].tsx]: Done! had to add an override for the button styles

File: pages/hold/request/[id].tsx (hunk 15 above)
Comment (dgcohen) [reply pages/hold/request/[id].tsx:225]: The 404 is just a redirect to the 404 page, we can make the error a 400 in the api route though.

**Review by charmingduchess — APPROVED** [review]

**Review by EdwinGuzman — APPROVED** [review]
Looks good! I think some updates will be needed once you have more parts implemented.

One big ask though, can you create a feature branch for the hold pages? Then point this pr at the feature branch. This way, `main` is not affected and you can work on the Hold pages separately, and it might help reduce merge conflicts and deploying the right code to production.
